internal/controller/apiextensions/usage/reconciler.go

```diff
@@ -316,6 +317,19 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 			}
 		}
 
+		if u.Spec.ReplayDeletion != nil && *u.Spec.ReplayDeletion && used.GetAnnotations() != nil && used.GetAnnotations()[usage.AnnotationKeyFirstDeletionAttempt] != "" {
+			// We have already recorded a deletion attempt and want to replay deletion, let's delete the used resource.
+			log.Debug("Replaying deletion of the used resource", "apiVersion", used.GetAPIVersion(), "kind", used.GetKind(), "name", used.GetName())
+			go func() {
+				// We do the deletion async and after some delay to make sure the usage is deleted before the
+				// deletion attempt. We remove the finalizer on this Usage right below, so, we know it will disappear
+				// very soon.
+				time.Sleep(2 * time.Second)
+				// We cannot use the context from the reconcile function since it will be cancelled after the reconciliation.
+				_ = r.client.Delete(context.Background(), used)
```

Comment:
Does this have to consider passing the right DeleteOptions for foreground propagationPolicy ?

What if the used resource is from a different composition than the Usage resource and the by/using resource and happens to be an XR composite from a claim with Foreground compositeDeletionPolicy.   In that case, letting the claim reconciler retrying the delete with appropriate deleteOptions seems to be the right thing ?

Reply:
> Does this have to consider passing the right DeleteOptions for foreground propagationPolicy ?

That's a good point - it would make sense to use the same DeleteOptions, and in my python implementation of this feature I was using Foreground deletion here by default (buried in a method default parameter so I wasn't thinking about it).  Theoretically we're just trying to wake up the GC process early, since it has already marked things for deletion and set finalizers I don't think the propagation policy matters here, but it is possible that it could make a difference.  Probably better to set it than not.  I think we could check for a `foregroundDeletion` finalizer on the resource and if it exists then retry with `Foreground` propagation policy otherwise let it default to `Background`.

I don't know if retrying with the Claim reconciler will affect the lower-level resources that GC is trying to delete.  The delete call on the Claim would need to trigger GC on the lower-level resources and I don't know if that happens or not.

Reply:
> Theoretically we're just trying to wake up the GC process early, since it has already marked things for deletion and set finalizers I don't think the propagation policy matters here, but it is possible that it could make a difference. Probably better to set it than not. I think we could check for a foregroundDeletion finalizer on the resource and if it exists then retry with Foreground propagation policy otherwise let it default to Background.

Did you mean checking for the foregroundDeletion finalizer presence on the parent of `used` resource (from its ownerReferences) ?  the used resource itself wouldn't have either `metadata.deletionTimestamp` or the foregroundDeletion finalizer set on it at this stage, as the presence of Usage would have rejected the previous deletion attempts.

> I don't know if retrying with the Claim reconciler will affect the lower-level resources that GC is trying to delete. The delete call on the Claim would need to trigger GC on the lower-level resources and I don't know if that happens or not.

If the used/`of` resource referred in the Usage happens to be a nested XR or MR composed as part of a composite (child resources with metadata.OwnerReferences), then only it is possible for the K8s GC Controller to be the one attempting delete, right ?

If the used/`of` resource referred in the Usage happens to be a parent XR(without any metadata.OwnerReferences set) from a different composition than the using/Usage itself,  the flow of deletion would be :
External Client issues delete on the Claim with Foreground compositeDeletionPolicy ->  Claim reconciler attempts to delete the XR with foreground propagation policy -> delete fails due to Usage ->  claim reconciler keeps attempting delete on XR with requeues.. there are no GC controller invoked deletes on such a parent XR, in this scenario..

Reply:
Good call @ravilr 👍 
I updated the PR to replay with the same propagation policy.